contracts/azorius/LinearERC721Voting.sol

```diff
@@ -0,0 +1,292 @@
+        uint32 proposalId = abi.decode(_data, (uint32));
+        uint32 _votingEndBlock = uint32(block.number) + votingPeriod;
+
+        proposalVotes[proposalId].votingEndBlock = _votingEndBlock;
+        proposalVotes[proposalId].votingStartBlock = uint32(block.number);
+
+        emit ProposalInitialized(proposalId, _votingEndBlock);
+    }
+
+    function addGovernanceToken(address _nftAddress, uint256 _weight, bool _isProposer) public onlyOwner {
+        IERC721 token = IERC721(_nftAddress);
+        if (!token.supportsInterface(0x80ac58cd))
+            revert InvalidTokenAddress(); 
+        
+        for (uint i = 0; i < governanceTokens.length; i++) {
+            if (address(governanceTokens[i].token) == _nftAddress) {
+                revert TokenAlreadySet();
+            }
+        }
+
+        GovernanceNFT memory governance = GovernanceNFT({
+            token: token,
+            weight: _weight,
+            isProposer: _isProposer
+        });
+
+        governanceTokens.push(governance);
+        emit GovernanceTokenAdded(_nftAddress);
+    }
+
+    function removeGovernanceToken(address _nftAddress) external onlyOwner {
+        if (governanceTokens.length == 0) revert LastGovernanceToken();
+
+        for (uint256 i = 0; i < governanceTokens.length; i++) {
+            if (address(governanceTokens[i].token) == _nftAddress) {
+                
+                uint256 last = governanceTokens.length - 1;
+                governanceTokens[i] = governanceTokens[last]; // move the last token into the position to remove
+                delete governanceTokens[last];                // delete the last token
+
+                emit GovernanceTokenRemoved(_nftAddress);
+                break;
+            }
+        }
+    }
+
+    /**
+     * Returns whether an NFT has already voted.
+     */
+    function hasVoted(uint32 _proposalId, address _tokenAddress, uint256 _tokenId) public view returns (bool) {
+        return proposalVotes[_proposalId].hasVoted[_tokenAddress][_tokenId];
+    }
+
+    /** @inheritdoc BaseStrategy*/
+    function isPassed(uint32 _proposalId) public view override returns (bool) {
+        return (
+            block.number > proposalVotes[_proposalId].votingEndBlock && // voting period has ended
+            quorumThreshold <= proposalVotes[_proposalId].yesVotes + proposalVotes[_proposalId].abstainVotes && // yes + abstain votes meets the quorum
+            meetsBasis(proposalVotes[_proposalId].yesVotes, proposalVotes[_proposalId].noVotes) // yes votes meets the basis
+        );
+    }
+
+    /** @inheritdoc BaseStrategy*/
+    function isProposer(address _address) public view override returns (bool) {
+        for (uint i = 0; i < governanceTokens.length; i++) {
+            GovernanceNFT memory governance = governanceTokens[i];
+            if (governance.isProposer && governance.token.balanceOf(_address) > 0) {
+                return true;
+            }
+        }
```

Comment:
This loop also could removed if a mapping is used for the governanceTokens data

Reply:
@tbwebb22 how can we do this here with just a mapping of governance tokens?  We need to know each token address to check if they have a balance